File: w3/dz_cpp/bcw4/bcw4/Pen&Paper/Paper.cpp
#include <iostream>
#include "Paper.h"

Paper::Paper(int maxSymbols) {
    this->maxSymbols = maxSymbols;
    this->symbols = 0;
    this->content = "";
}

Paper::~Paper() {}

int Paper::getMaxSymbols() const {
    return this->maxSymbols;
}

int Paper::getSymbols() const {
    return this->symbols;
}

void Paper::addContent(const std::string& message) {
    if ( this->symbols == this->maxSymbols ) {
        throw OutOfSpace();
    }
    int maxInputSymbols = this->maxSymbols - this->symbols;

    if ( maxInputSymbols < message.size() ) {
        this->symbols = this->maxSymbols;
        this->content += message.substr(0, maxInputSymbols);
        return;
    }
    this->symbols += message.size();
    this->content += message;
}

void Paper::show() const {
    std::cout << "Paper: " << this->symbols << "/" << this->maxSymbols << "\n";
    std::cout << this->content << std::endl;
}

File: w3/dz_java/src/PenPaper/Paper.java
public class Paper {
    private int maxSymbols;
    private int symbols;
    private StringBuffer content;

    public Paper() {
        this.maxSymbols = 4096;
        this.symbols = 0;
        content = new StringBuffer();
    }

    public Paper(int maxSymbols) {
        this.maxSymbols = maxSymbols;
        this.symbols = 0;
        content = new StringBuffer();
    }

    public int getMaxSymbols() {
        return maxSymbols;
    }

    public int getSymbols() {
        return symbols;
    }

    public StringBuffer getContent() {
        return content;
    }

    public void addContent(String message) throws OutOfSpace {
        int maxInputSymbols = this.maxSymbols - this.symbols;

        if (this.symbols == this.maxSymbols ) {
            throw new OutOfSpace();
        }

        if ( maxInputSymbols < message.length() ) {
            this.symbols = this.maxSymbols;
            this.content.append(message.substring(0, maxInputSymbols));
            return;
        }
        this.symbols += message.length();
        this.content.append(message);
    }

    public void show() {
        System.out.println(this.getContent());
    }
}
